File: backend/engine/thinking.py
import re


def strip_thinking(text: str) -> tuple[str, str]:
    """
    Strip thinking/reasoning blocks from model output.
    Returns (cleaned_output, thinking_content).
    Handles: <think>...</think>, plain-text "Thinking Process:" blocks.
    """
    thinking = ""

    # 1. <think>...</think> tags (Qwen3.5, DeepSeek R1, etc.)
    think_match = re.search(r'<think>(.*?)</think>', text, flags=re.DOTALL)
    if think_match:
        thinking = think_match.group(1)
        text = re.sub(r'<think>.*?</think>\s*', '', text, flags=re.DOTALL)

    # 2. Unclosed <think> (model ran out of tokens mid-thinking)
    elif '<think>' in text and '</think>' not in text:
        idx = text.index('<think>')
        thinking = text[idx + 7:]
        text = text[:idx]

    # 3. "Thinking Process:\n..." blocks (some Qwen versions)
    tp_match = re.match(r'Thinking Process:\s*\n(.*?)(?=\n\n[A-Z`#]|\Z)', text, flags=re.DOTALL)
    if tp_match and not thinking:
        thinking = tp_match.group(1)
        text = text[tp_match.end():]

    return text.strip(), thinking.strip()

File: src/main/engine/thinking.ts
export interface ThinkingExtraction {
  cleaned: string;
  thinking: string;
  truncated: boolean;
}

export function extractThinking(text: string): ThinkingExtraction {
  let cleaned = text;
  let thinking = '';
  let truncated = false;

  const thinkMatch = text.match(/<think>([\s\S]*?)<\/think>/);
  if (thinkMatch) {
    thinking = thinkMatch[1] ?? '';
    cleaned = text.replace(/<think>[\s\S]*?<\/think>\s*/s, '');
    return {
      cleaned: cleaned.trim(),
      thinking: thinking.trim(),
      truncated: false,
    };
  }

  if (text.includes('<think>') && !text.includes('</think>')) {
    const index = text.indexOf('<think>');
    thinking = text.slice(index + 7);
    cleaned = text.slice(0, index);
    truncated = true;
    return {
      cleaned: cleaned.trim(),
      thinking: thinking.trim(),
      truncated,
    };
  }

  if (!text.includes('<think>') && text.includes('</think>')) {
    const index = text.indexOf('</think>');
    thinking = text.slice(0, index);
    cleaned = text.slice(index + 8);
    return {
      cleaned: cleaned.trim(),
      thinking: thinking.trim(),
      truncated: false,
    };
  }

  const thinkingProcessMatch = text.match(
    /^Thinking Process:\s*\n(.*?)(?=\n\n[A-Z`#]|\Z)/s,
  );
  if (thinkingProcessMatch) {
    thinking = thinkingProcessMatch[1] ?? '';
    cleaned = text.slice(thinkingProcessMatch[0].length);
    return {
      cleaned: cleaned.trim(),
      thinking: thinking.trim(),
      truncated: false,
    };
  }

  return {
    cleaned: text.trim(),
    thinking: '',
    truncated: false,
  };
}

export function stripThinking(text: string): [string, string] {
  const extracted = extractThinking(text);
  return [extracted.cleaned, extracted.thinking];
}
